compiler-lib/src/optimization/control_flow.rs

```diff
@@ -0,0 +1,599 @@
+                let mut label = default_label(node);
+
+                if node == &Node::Block(block) {
+                    label = label
+                        .style(Style::Filled)
+                        .fillcolor(X11Color::Red)
+                        .fontcolor(X11Color::White);
+                }
+
+                label
+            }
+        );
+
+        if self.done.contains(&block) {
+            self.done.remove(&block);
+        }
+
+        if !self.worklist.contains(&block) {
+            self.worklist.push_back(block);
+        }
+    }
+
+    /// Try to replace an unnecessary conditional jump with an
+    /// unconditional jump. A conditional jump is unnecessary if:
+    ///
+    /// 1. true projection and false projection jump to the same
+    ///    target block [using input array element `i` and `j`].
+    /// 2. For all phi nodes in the target block, the `i`-th and `j`-th
+    ///    value have to be identical
+    /// 3. The true and false projection have to be the successors
+    ///    of the same cond node. This is guaranteed in our implementation
+    ///    since we do not use switch nodes.
+    ///
+    ///
+    /// This is a sketch of the structure we are trying to match:
+    ///
+    /// ```text
+    ///     Predecessor Block
+    ///      |           |
+    ///     Cond       Other Nodes are
+    ///    /    \      possible/allowed
+    /// Proj   Proj
+    /// True   False
+    ///   |     |
+    /// in[i]   in[j]
+    ///   |     |
+    /// Current Block --- Other Nodes are possible/allowed
+    ///  |      |
+    ///  Phi 0  Phi N # forall phi in Current Block: phi[i] == phi[j]
+    /// ```
+    fn visit_proj(&mut self, current: Node) -> bool {
+        if let Some(unnecessary_cond) = self.match_unnecessary_cond(current) {
+            // To replace a conditional jump with an unconditional jump, we:
+            //
+            // 1. Reuse the `i`-th element of the input array and phi nodes
+            //    for the jump node. There is no need to adapt the phi nodes
+            //    since the value at `i` is still correct.
+            // 2. Remove index `j` from the input array and all phi nodes.
+            //    (Since `j` is not guranteed to be the last input element,
+            //    this means we have to "compact" the arrays by moving
+            //    indices `x > j` to `x - 1`, but thats done by the vector
+            //    implementation for us.)
+            let jmp = unnecessary_cond.cond.block().new_jmp();
+            let mut target_block_preds =
+                unnecessary_cond.target_block.in_nodes().collect::<Vec<_>>();
+            target_block_preds[unnecessary_cond.proj_current.0] = Node::Jmp(jmp);
+
+            // remove index of second projection, called step (2) above
+            let idx = unnecessary_cond.proj_other.0;
+            for phi in unnecessary_cond.target_block.phis() {
+                let mut phi_preds = phi.phi_preds().collect::<Vec<_>>();
```

Comment:
Ich hab in solchen Fällen immer `phi_preds: Vec<_> = (...).collect()` gemacht...